File: leetcode/python/787_cheapest_flights_within_k_stops.py
import heapq
import sys


class Solution:
    def findCheapestPrice(self, n: int, flights: list[list[int]], src: int, dst: int, k: int) -> int:
        g = self.buildGraph(n, flights)
        heap = [(0, 0, src)]
        visited = [k+1] * n
        while heap:
            cost, depth, node = heapq.heappop(heap)
            if node == dst:
                return cost
            if depth >= visited[node]:
                continue
            visited[node] = depth
            for v, c in g[node]:
                heapq.heappush(heap, (cost + c, depth + 1, v))
        return -1
        
    def buildGraph(self, n: int, flights: list[list[int]]) -> list[list[tuple[int,int]]]:
        g = [[] for _ in range(n)]
        for u, v, cost in flights:
            g[u].append((v, cost))
        return g
    
s = Solution()
print(s.findCheapestPrice(n = 4, flights = [[0,1,100],[1,2,100],[2,0,100],[1,3,600],[2,3,200]], src = 0, dst = 3, k = 1))
print(s.findCheapestPrice(n = 3, flights = [[0,1,100],[1,2,100],[0,2,500]], src = 0, dst = 2, k = 1))

File: leetcode/rust/src/787_cheapest_flights_within_K_stops.rs
impl Solution {
    pub fn find_cheapest_price(n: i32, flights: Vec<Vec<i32>>, src: i32, dst: i32, k: i32) -> i32 {
        // bellman-ford
        let n = n as usize;
        let mut dist = vec![std::i32::MAX; n];
        dist[src as usize] = 0;

        for _ in 0..k + 1 {
            let mut tmp = dist.clone();
            for flight in &flights {
                let u = flight[0] as usize;
                let v = flight[1] as usize;
                let w = flight[2];
                if dist[u] != std::i32::MAX && dist[u] + w < tmp[v] {
                    tmp[v] = dist[u] + w;
                }
            }
            dist = tmp;
        }

        if dist[dst as usize] == std::i32::MAX {
            -1
        } else {
            dist[dst as usize]
        }

    }
}